### PR TITLE
Traefik: 1.7.14 -> 2.2.0

### LINKED ISSUE #76518: Traefik 2.0
**Project description**
Traefik is a microservice oriented reverse proxy written in Go.
It configures itself by container labels from docker/other provider or  from text files.
It does TLS termination and can fetch certs from let's encrypt automatically.

A nixOS module to add configuration to the file backend (similar to the nginx ones) seems easy to implement.

Traefik 1.0 is currently packaged ([module](https://github.com/NixOS/nixpkgs/blob/dd0a47e7ae2f6f33298c61f75aff2652e0a2f6d6/nixos/modules/services/web-servers/traefik.nix), [package](https://github.com/NixOS/nixpkgs/blob/69d58ee24506e2b3b5aff347ae3b0791110340ec/pkgs/servers/traefik/default.nix))

Traefik 2.0 is NOT backwards compatible

A package for 2.0 (not source built) has been written up by @ljani [here](https://github.com/NixOS/nixpkgs/issues/41378#issuecomment-541453536)

**Metadata**

* homepage URL: https://docs.traefik.io/
* source URL: https://github.com/containous/traefik
* license: MIT
* platforms: linux, darwin, freebsd, openbsd, windows

### PR DESCRIPTION
**Pull request by jokogr**
###### Motivation for this change

I was experimenting with Traefik and thought to update the package to the latest v2 one.

In the mean time, `buildGoPackage` was switched to `buildGoModule` and the respective NixOS module has been adapted to the new configuration requirements (static and dynamic, routers, middlewares, services and so on).

Finally, I have removed `hamhut1066` from the maintainers list, as this user is no longer in GitHub.

Fixes #76518.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @vdemeester

### CODE REVIEW

**Review by Br1ght0ne — APPROVED** [review]
diff for package LGTM
didn't check module
executable shows usage

```
[5 built, 5 copied (432.5 MiB), 104.1 MiB DL]
https://github.com/NixOS/nixpkgs/pull/76723
1 package built:
traefik
```

**Comment by dali99** [comment]
Considering that the configuration isn't backwards compatible, should this maybe go into a separate package/module?

traefik 1.7 is still supported upstream until september 2020, and this kind of change should definetly not be backported to 19.09

having this be the only option on master and 20.03 seems reasonable

**Comment by jokogr** [comment]
> Considering that the configuration isn't backwards compatible, should this maybe go into a separate package/module?
> 
> traefik 1.7 is still supported upstream until september 2020, and this kind of change should definetly not be backported to 19.09
> 
> having this be the only option on master and 20.03 seems reasonable
> 
> Otherwise LGTM!

Thanks for having a look on this PR. I do not understand your initial question, as in whether to have a separate package / module. As you point out after the question and I agree with this, we could maintain Traefik v1.7 on 19.09 only and v2.1 and later on 20.03 and master. Why should we have additional maintenance effort on master?

@vdemeester what's your view on this?

**Comment by dali99** [comment]
Yeah I'm just not sure on what the general policy is on changes like this. It's kind of two entirely different pieces of software, and 20.09 is the first release we can do where 1.7 is no longer supported upstream. 

I'm personally okay with 2.1 being the only thing available on master and 20.03, but I just wanted to put the info here so other's could chime in.

**Comment by dali99** [comment]
@vdemeester 
@kalbasit 

Friendly pings :D (3 weeks) has this been forgotten?

File: pkgs/servers/traefik/default.nix
Comment (kalbasit) [inline pkgs/servers/traefik/default.nix]:
```suggestion
```
This is unused by `buildGoModule`.

File: pkgs/servers/traefik/default.nix
Comment (kalbasit) [inline pkgs/servers/traefik/default.nix]: Please move these out of the `preBuild` phase. Example :point_down: https://github.com/NixOS/nixpkgs/blob/75a9868dbe66bff430d3e1c01b5067a1c48053c3/pkgs/applications/networking/cluster/helm/default.nix#L16

File: nixos/modules/services/web-servers/traefik.nix
Comment (kalbasit) [inline nixos/modules/services/web-servers/traefik.nix]: Don't you need to set `entryPoints.web.address` like in the example case?

File: pkgs/servers/traefik/default.nix
Comment (jokogr) [reply pkgs/servers/traefik/default.nix]: I had them there because of the `CODENAME` variable, which gets evaluated at the `preBuild` phase. Are you sure that `buildFlagsArray` is going to be evaluated properly?

File: nixos/modules/services/web-servers/traefik.nix
Comment (jokogr) [reply nixos/modules/services/web-servers/traefik.nix]: To be honest, I didn't see anything `defaultEntryPoints` on the [official documentation site](https://docs.traefik.io/), so I skipped this.

This [issue](https://github.com/containous/traefik/issues/5184) on the official traefik GitHub repo supports this.

**Comment by jokogr** [comment]
Thanks @kalbasit for the review and sorry for the wait! I have made the following changes:

* Refactored to current master.
* Updated to Traefik v2.1.4.
* Removed `goPackagePath` as suggested.
* Made a new commit in the relevant NixOS service to make configuration deep mergeable as in 00eb055162a52a2d9f42ba6c0eeec892c83bfa58 (thanks @Mic92 for the hint)

File: pkgs/servers/traefik/default.nix
Comment (rnhmjoj) [reply pkgs/servers/traefik/default.nix]: I'll side with @jokogr: `buildFlagsArray` can't be used here. The best we can do is use `makeFlagsArray`, which is what the nixpkgs manual recommends:

```suggestion
    makeFlagsArray+=("-ldflags=\
      -X github.com/containous/traefik/version.Version=${version} \
      -X github.com/containous/traefik/version.Codename=$CODENAME")

```

**Comment by jokogr** [comment]
Thanks @rnhmjoj for the review! I have made the following changes:

* Rebased to current master.
* Updated traefik to v2.1.6.
* Switched to `makeFlagsArray` as suggested by @rnhmjoj (sorry, rebasing took away @kalbasit's and @rnhmjoj's comments).

**Comment by rnhmjoj** [comment]
So, the changes look good and I tested the package update.
I would be more confident to merge this if someone would review the module: it's a substantial change. Also it would be good to have a NixOS test, even a simple one.

File: nixos/modules/services/web-servers/traefik.nix
Comment (Mic92) [inline nixos/modules/services/web-servers/traefik.nix]: Is the static/dynamic configuration a concept that traefic has or is it something that NixOS introduces? If it is the latter one, it should be mentioned in the documentation.

Comment (dali99) [reply nixos/modules/services/web-servers/traefik.nix]: Static vs Dynamic configuration is a concept Traefik has. Static being that specific config file. And dynamic being for example a label on a container.

File: pkgs/servers/traefik/default.nix
Comment (Mic92) [inline pkgs/servers/traefik/default.nix]:
```suggestion
  nativeBuildInputs = [ go-bindata ];
```

**Comment by dali99** [comment]
Changed the module to use attrsOf instead of lazyAttrsOf as a test on 19.09

```
hash mismatch in fixed-output derivation '/nix/store/hb3xrk6vzszz87vz6z0sk1zdcjssf0r3-traefik-2.1.6-go-modules':
  wanted: sha256:1mdzjc9v54b0gc8f97gpdgym1dxhyc1p1qivgfvjraj3vgfc7i5g                          
  got:    sha256:1chd174wplr0sqyij4z05kpnhs81sd3xc7jjlb2jlxmbnsf9kay5
```

changing the hash:

```
build github.com/containous/traefik/v2/internal: cannot load crypto/ed25519: cannot find module providing package crypto/ed25519
generate.go:5: running "go": exit status 1
```

I dont actually know the difference between lazyAttrsOf and attrsOf semantically, so that could be the reason its failing like it is, in that case disregard this comment.

20.03 should be out soon, Ill try build it ontop of that if (it doesnt get merged into it) when that happens.

**Comment by jokogr** [comment]
* Rebased to current master.
* Updated traefik to v2.2.0.
* Switched from `buildInputs` to `nativeBuildInputs` as suggested by @Mic92.

@dali99 please, do not use words like `Package should build anyways, but it doesnt`, it gives a wrong impression of this PR.

Traefik as of now requires Go 1.14 (which is the default version currently in unstable) and NixOS 19.09 **merely** supports 1.13 and it is not working as in unstable, as you have noticed.

It happens to use this package (and NixOS module) on my servers running NixOS 19.09 and it works as intended after adding Go 1.14 in a package overlay:

```
  buildGo114Module = super.callPackage ./go-modules_generic.nix {
    go = self.buildPackages.go_1_14;
  };
  go_1_14 = super.callPackage ./go_1.14 {
    inherit (super.darwin.apple_sdk.frameworks) Security Foundation;
  };
  traefik = super.callPackage ./traefik {
    buildGoModule = self.buildGo114Module;
  };
```

The versions and the files I have copied there is left to the readers as an exercise. As I have mentioned before, I would like to focus my efforts on unstable to speed up the update process.

@rnhmjoj I will try to create a NixOS test for traefik today or tomorrow.

**Comment by dali99** [comment]
I'm sorry I was just trying to help :disappointed:

**Comment by jokogr** [comment]
> I'm sorry I was just trying to help disappointed
> 
> Good that I was just personally failing to build it though!

No worries, I know you meant well, but I also do not want reviewers to have an excuse for not reviewing this :smiley:

* Added a test to check the Docker and file providers.

File: pkgs/servers/traefik/default.nix
Comment (Mic92) [inline pkgs/servers/traefik/default.nix:17]:
```suggestion
  nativeBuildInputs = [ go-bindata ];
  
  passthru.tests = { inherit (nixosTests) traefik; };
```

**Comment by jokogr** [comment]
* Add `passthru.tests` as suggested by @Mic92.

**Comment by jokogr** [comment]
* Added `nixosTests` in the argument list. :roll_eyes:

**Comment by jokogr** [comment]
* Set a limit on x86_64-linux for the test, as it relies on docker-containers, whose tests have the same limit.

File: nixos/modules/services/web-servers/traefik.nix
Comment (rnhmjoj) [inline nixos/modules/services/web-servers/traefik.nix:31]: Is there a reason you couldn't use `lib.recursiveUpdate`? It seems to me it does the same thing, at least in this case.

File: nixos/modules/services/web-servers/traefik.nix
Comment (rnhmjoj) [inline nixos/modules/services/web-servers/traefik.nix:52]: Can you break this line? It's really looong.

**Review by rnhmjoj — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-servers/traefik.nix
Comment (jokogr) [reply nixos/modules/services/web-servers/traefik.nix]: I wasn't aware of it, it looks indeed similar, thanks!

File: nixos/modules/services/web-servers/traefik.nix
Comment (jokogr) [reply nixos/modules/services/web-servers/traefik.nix]: Agreed, I've used nixfmt to format the whole file.

**Comment by jokogr** [comment]
* Rebased to current master.
* Dropped `recursiveMerge` as requested by @rnhmjoj.
* Formatted the package and the service nix files using nixfmt to address a code change request by @rnhmjoj.

**Review by rnhmjoj — APPROVED** [review]
Well, I don't have anything else to add: it looks pretty good to me.
I double checked the package and the moduble, everything builds and ofBorgs is happy.
Thank you for adding a test: it will make future changes easier to review.

I'll wait a bit for the other reviewers if the have anything else to say, before merging.

**Pull request merged**
